src/redactableProcessor.js

```diff
@@ -43,23 +45,61 @@ module.exports = class RedactableProcessor {
       .stringify(sourceTree);
   }
 
-  sourceAndRedactedToMergedSyntaxTree(source, redacted) {
-    const sourceTree = this.sourceToRedactedSyntaxTree(source);
-    const redactedTree = this.getProcessor()
+  /* This function checks that
+   * 1. Redactions in the sourceTree == restorations performed
+   * 2. There are no redactions that weren't performed
+   * 3. There was no additions redactable content in the mergedTree
+   * */
+  checkRestorationNodes(sourceTree, mergedTree) {
```

Comment:
is it necessary to check all these things? Seems like we could just check for `redaction` and `unrestored` nodes in the mergedTree, which would have the added benefit of not adding the `restored` property to MDAST nodes

Reply:
I think the check that we restore the right number of redactions is necessary to detect missing redactions. I can't think of any other way to check this except to keep track of `restored` nodes.

Reply:
Ahh, of course.

I mostly bring this up because I really don't like adding a property to [MDAST nodes](https://github.com/syntax-tree/mdast). It violates a contract and introduces an element of state to the syntax tree that really shouldn't be there.

What if we include the redacted tree in this check? Something like

```js
const selectAll = require('unist-util-select').selectAll;
const select = require('unist-util-select').select

const sourceRedactions = selectAll('redaction', sourceTree);
const redactedRedactions = selectAll('redaction', redactedTree);

// If source and redacted have mismatching number of redactions, this
// means a redaction was either added or removed in the redacted text
if (sourceRedactions.length !== redactedRedactions.length) {
  return false;
}

// Even if both strings have the same total number of redactions, they
// may have mismatching redaction indexes or other errors that prevent
// perfect restoration
if (select('unrestored', mergedTree)) {
  return false;
}

return true;
```

Reply:
It occurs to me this would require us to be able to parse redacted text into a syntax tree, which we can't currently do. But I think that's a good thing! Right now, restoration happens as part of the parsing process, which is really strange; specifically, it's a thing that requires a syntax tree and some text, and produces a second syntax tree. What if we instead added the ability to parse redacted text to a syntax tree, and then made restoration be a process that takes two syntax trees and produces a third? That seems _much_ cleaner